File: backend/controllers/offeringsController.js
const db = require('../config/db');
const jwt = require('jsonwebtoken');
const offeringsModel = require('../models/offeringsModel');
const userModel = require('../models/usersModel');

exports.getOfferings = async (req, res, next) => {
    try {
        let offers = await offeringsModel.fetchOffers();
        return res.json(offers);
    } catch (err) {
        next(err);
    }
}

exports.getOfferingsWithFilters = async (req, res, next) => {
    try{
        const filters = req.body.filters || {};
        let offers = await offeringsModel.fetchFilteredOffers(filters);
        return res.json(offers);
    }
    catch(err){
        next(err);
    }
}

exports.getOfferingAvailableTime = async (req, res, next) => {
    try{
        const offerId = req.params.offeringId;

        if(offerId === undefined || offerId === null){
            return res.status(400).json({message:"No offering Id provided"});
        }

        const offeringInfo = await db.query(
            `Select start_at, end_at From offerings o, time_slots t
                Where o.id = $1 and o.provider_id = t.provider_id
                Order By start_at ASC`
        , [offerId]);


        let busyTimes = [];
        let freeTimes = [];

        for(let i = 0; i < offeringInfo.rows.length; i++){
            busyTimes.push([new Date(offeringInfo.rows[i].start_at), new Date(offeringInfo.rows[i].end_at)]);
        }
        

        let future = false;
        let lastEnd = Date.now();
        for(let i = 0; i < busyTimes.length; i++){
            if(future){
                if(i > 0){
                    freeTimes.push([busyTimes[i-1][1].getTime(), busyTimes[i][0].getTime()]);
                    lastEnd = busyTimes[i][0].getTime();
                }
            }
            else{
                if(Date.now() < busyTimes[i][0].getTime()){
                    future = true;
                    freeTimes.push([Date.now(), busyTimes[i][0].getTime()]);
                    lastEnd = busyTimes[i][0].now();
                }
                else if(Date.now() >= busyTimes[i][0].getTime() && Date.now() <= busyTimes[i][1].getTime()){
                    future = true;
                }
            }
        }
        freeTimes.push([lastEnd, Date.now() + 100*365*24*60*60*1000]);

        return res.json(freeTimes);
    }
    catch(err){
        next(err);
    }
}

exports.getProviderOffers = async (req, res, next) => {
    try {

        const authHeader = req.headers['authorization'];
        
        if (!authHeader || !authHeader.startsWith('Bearer '))
            return res.status(401).json({ error: 'No token provided' });

        const token = authHeader.split(' ')[1];

        // Verify JWT
        const payload = jwt.verify(token, process.env.JWT_SECRET);
        req.user = payload; // attach user_id
        const { user_id } = payload;

        if(!userModel.isAProvider(user_id)){
            return res.status(401).json({message : "You are not a provider"});
        }

        let offers = await offeringsModel.fetchProviderOffers(user_id);
        return res.json(offers);
    } catch (err) {
        next(err);
    }
}

exports.createProviderOffer = async (req, res, next) => {
    try {

        const authHeader = req.headers['authorization'];
        
        if (!authHeader || !authHeader.startsWith('Bearer '))
            return res.status(401).json({ error: 'No token provided' });

        const token = authHeader.split(' ')[1];

        // Verify JWT
        const payload = jwt.verify(token, process.env.JWT_SECRET);
        // req.user = payload; // attach user_id
        const { user_id } = payload;

        let isProvider = await userModel.isAProvider(user_id);
        if(!isProvider){
            return res.status(401).json({message : "You are not a provider"});
        }

        const { offer } = req.body;

        if(!offer.service_id || !offer.title || offer.rate === null || offer.rate === undefined || 
            !offer.curr || offer.active === null || offer.active === undefined){
                return res.status(401).json({message : "Please fill all required fields"});
        }

        const resultt = await offeringsModel.createOffer(user_id, offer);
        if(!resultt){
            return res.status(401).json({message : "Failed to create offer"});
        }

        return res.json({success : true, offerId : resultt});
    } catch (err) {
        next(err);
    }
}

exports.editProviderOffer = async (req, res, next) => {
    try{
        const authHeader = req.headers['authorization'];
        
        if (!authHeader || !authHeader.startsWith('Bearer '))
            return res.status(401).json({ error: 'No token provided' });

        const token = authHeader.split(' ')[1];

        // Verify JWT
        const payload = jwt.verify(token, process.env.JWT_SECRET);
        // req.user = payload; // attach user_id
        const { user_id } = payload;

        const isProvider = await userModel.isAProvider(user_id);
        if(!isProvider){
            return res.status(401).json({message : "You are not a provider"});
        }

        const offeringId = req.params.offeringId;
        if(!offeringId){
            return res.status(401).json({message : "Invalid Offering Id"});
        }

        const isOfferOwner = await userModel.isOfferOwner(user_id, offeringId);
        if(!isOfferOwner){
            return res.status(401).json({message : "You are not the owner of this offering"});
        }

        const { offer } = req.body;

        if(!offer.title || offer.rate == null || offer.rate == undefined ||
            !offer.curr || offer.active == null || offer.active == undefined || !offer.service_id){
                return res.status(401).json({message : "Please fill all required fields"});
        }

        const patchResult = await offeringsModel.updateOffer(offeringId, offer);
        if(!patchResult){
            return res.status(401).json({message : "Failed to update offer data"});
        }

        return res.json({success: true});
    }
    catch(err){
        next(err);
    }
}

exports.deleteProviderOffer = async (req, res, next) => { 
    try{
        const authHeader = req.headers['authorization'];
        
        if (!authHeader || !authHeader.startsWith('Bearer '))
            return res.status(401).json({ error: 'No token provided' });

        const token = authHeader.split(' ')[1];

        // Verify JWT
        const payload = jwt.verify(token, process.env.JWT_SECRET);
        // req.user = payload; // attach user_id
        const { user_id } = payload;

        const isProvider = await userModel.isAProvider(user_id);
        if(!isProvider){
            return res.status(401).json({message : "You are not a provider"});
        }

        const offeringId = req.params.offeringId;
        if(!offeringId){
            return res.status(401).json({message : "Invalid Offering Id"});
        }

        const isOfferOwner = await userModel.isOfferOwner(user_id, offeringId);
        if(!isOfferOwner){
            return res.status(401).json({message : "You are not the owner of this offering"});
        }

        const deleteResult = await offeringsModel.deleteOffer(offeringId);
        if(!deleteResult){
            return res.status(401).json({message : "Failed to delete offer"});
        }

        return res.json({success: true});
    }
    catch(err){
        next(err);
    }
}

File: backend-csharp/Controllers/OfferingsController.cs
using HomeServicesPlatform.Data;
using HomeServicesPlatform.Filters;
using HomeServicesPlatform.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace HomeServicesPlatform.Controllers
{
    [ApiController]
    [Route("api/v1/offerings")]
    public class OfferingsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public OfferingsController(ApplicationDbContext context)
        {
            _context = context;
        }

        [HttpGet]
        public async Task<IActionResult> GetOfferings()
        {
            var offers = await _context.offerings
                .Include(o => o.Provider)
                    .ThenInclude(p => p.User)
                .Include(o => o.Provider)
                    .ThenInclude(p => p.Address)
                .Include(o => o.Service)
                .Where(o => o.Active)
                .Select(o => new
                {
                    offerId = o.Id,
                    providerName = o.Provider.User.Name,
                    serviceName = o.Service.Name,
                    providerCountry = o.Provider.Address.Country,
                    providerCity = o.Provider.Address.City,
                    offerTitle = o.Title,
                    hourlyRate = o.Rate,
                    currency = o.Curr
                })
                .ToListAsync();

            return Ok(offers);
        }

        [HttpPost]
        public async Task<IActionResult> GetOfferingsWithFilters([FromBody] FilterRequest filters)
        {
            var query = _context.offerings
                .Include(o => o.Provider)
                    .ThenInclude(p => p.User)
                .Include(o => o.Provider)
                    .ThenInclude(p => p.Address)
                .Include(o => o.Service)
                .Where(o => o.Active && o.Provider.User.Role == "provider");

            if (filters.Job != null && filters.Job.Count > 0)
            {
                var jobs = filters.Job.Select(j => j.ToLower()).ToList();
                query = query.Where(o => jobs.Contains(o.Service.Name.ToLower()));
            }

            if (filters.Rate != null)
            {
                if (filters.Rate.Min.HasValue)
                {
                    query = query.Where(o => o.Rate >= filters.Rate.Min.Value);
                }
                if (filters.Rate.Max.HasValue)
                {
                    query = query.Where(o => o.Rate <= filters.Rate.Max.Value);
                }
            }

            if (!string.IsNullOrEmpty(filters.Country))
            {
                query = query.Where(o => o.Provider.Address.Country.ToLower() == filters.Country.ToLower());
            }

            if (filters.Cities != null && filters.Cities.Count > 0)
            {
                var cities = filters.Cities.Select(c => c.ToLower()).ToList();
                query = query.Where(o => cities.Contains(o.Provider.Address.City.ToLower()));
            }

            var offers = await query.Select(o => new
            {
                offerId = o.Id,
                providerName = o.Provider.User.Name,
                serviceName = o.Service.Name,
                providerCountry = o.Provider.Address.Country,
                providerCity = o.Provider.Address.City,
                offerTitle = o.Title,
                hourlyRate = o.Rate,
                currency = o.Curr
            })
            .ToListAsync();

            return Ok(offers);
        }

        [HttpGet("available-time/{offeringId}")]
        public async Task<IActionResult> GetOfferingAvailableTime(int offeringId)
        {
            var offering = await _context.offerings
                .Include(o => o.Provider)
                .FirstOrDefaultAsync(o => o.Id == offeringId);

            if (offering == null)
            {
                return BadRequest(new { message = "No offering Id provided" });
            }

            var busyTimes = await _context.time_slots
                .Where(t => t.ProviderId == offering.ProviderId)
                .OrderBy(t => t.StartAt)
                .Select(t => new
                {
                    start = t.StartAt.Ticks,
                    end = t.EndAt.Ticks
                })
                .ToListAsync();

            var availableSlots = new List<List<long>>();
            var now = DateTime.UtcNow.Ticks;
            var oneHundredYearsLater = DateTime.UtcNow.AddYears(100).Ticks;

            for (int i = 0; i < busyTimes.Count; i++)
            {
                if (i == 0)
                {
                    if (now < busyTimes[i].start)
                    {
                        availableSlots.Add(new List<long> { now, busyTimes[i].start });
                    }
                }
                else
                {
                    if (now < busyTimes[i - 1].end)
                    {
                        availableSlots.Add(new List<long> { busyTimes[i - 1].end, busyTimes[i].start });
                    }
                    else if (now < busyTimes[i].start)
                    {
                        availableSlots.Add(new List<long> { now, busyTimes[i].start });
                    }
                }
            }

            if (busyTimes.Count > 0 && now < busyTimes[busyTimes.Count - 1].end)
            {
                availableSlots.Add(new List<long> { busyTimes[busyTimes.Count - 1].end, oneHundredYearsLater });
            }
            else
            {
                availableSlots.Add(new List<long> { now, oneHundredYearsLater });
            }

            return Ok(availableSlots);
        }

        [HttpGet("me")]
        [ServiceFilter(typeof(AuthFilter))]
        public async Task<IActionResult> GetProviderOffers()
        {
            var userId = (int)HttpContext.Items["UserId"]!;

            var provider = await _context.providers.FirstOrDefaultAsync(p => p.UserId == userId);
            if (provider == null)
            {
                return Unauthorized(new { message = "You are not a provider" });
            }

            var offers = await _context.offerings
                .Include(o => o.Provider)
                    .ThenInclude(p => p.User)
                .Include(o => o.Provider)
                    .ThenInclude(p => p.Address)
                .Include(o => o.Service)
                .Where(o => o.Provider.UserId == userId)
                .Select(o => new
                {
                    offerId = o.Id,
                    providerName = o.Provider.User.Name,
                    serviceName = o.Service.Name,
                    providerCountry = o.Provider.Address.Country,
                    providerCity = o.Provider.Address.City,
                    offerTitle = o.Title,
                    hourlyRate = o.Rate,
                    currency = o.Curr,
                    active = o.Active
                })
                .ToListAsync();

            return Ok(offers);
        }

        [HttpPost("me")]
        [ServiceFilter(typeof(AuthFilter))]
        public async Task<IActionResult> CreateProviderOffer([FromBody] CreateOfferRequest request)
        {
            var userId = (int)HttpContext.Items["UserId"]!;

            var provider = await _context.providers.FirstOrDefaultAsync(p => p.UserId == userId);
            if (provider == null)
            {
                return Unauthorized(new { message = "You are not a provider" });
            }

            if (request.Offer.ServiceId == 0 || string.IsNullOrEmpty(request.Offer.Title) || 
                request.Offer.Rate == null || string.IsNullOrEmpty(request.Offer.Curr) || 
                request.Offer.Active == null)
            {
                return BadRequest(new { message = "Please fill all required fields" });
            }

            var newOffer = new Offering
            {
                ProviderId = provider.Id,
                ServiceId = request.Offer.ServiceId,
                Title = request.Offer.Title,
                Rate = request.Offer.Rate.Value,
                Curr = request.Offer.Curr,
                Active = request.Offer.Active.Value
            };

            _context.offerings.Add(newOffer);
            await _context.SaveChangesAsync();

            return Ok(new { success = true, offerId = newOffer.Id });
        }

        [HttpPatch("{offeringId}")]
        [ServiceFilter(typeof(AuthFilter))]
        public async Task<IActionResult> EditProviderOffer(int offeringId, [FromBody] UpdateOfferRequest request)
        {
            var userId = (int)HttpContext.Items["UserId"]!;

            var provider = await _context.providers.FirstOrDefaultAsync(p => p.UserId == userId);
            if (provider == null)
            {
                return Unauthorized(new { message = "You are not a provider" });
            }

            var offering = await _context.offerings
                .Include(o => o.Provider)
                .FirstOrDefaultAsync(o => o.Id == offeringId);

            if (offering == null)
            {
                return BadRequest(new { message = "Invalid Offering Id" });
            }

            if (offering.Provider.UserId != userId)
            {
                return Unauthorized(new { message = "You are not the owner of this offering" });
            }

            if (string.IsNullOrEmpty(request.Offer.Title) || request.Offer.Rate == null || 
                string.IsNullOrEmpty(request.Offer.Curr) || request.Offer.Active == null || 
                request.Offer.ServiceId == 0)
            {
                return BadRequest(new { message = "Please fill all required fields" });
            }

            offering.ServiceId = request.Offer.ServiceId;
            offering.Title = request.Offer.Title;
            offering.Rate = request.Offer.Rate.Value;
            offering.Curr = request.Offer.Curr;
            offering.Active = request.Offer.Active.Value;

            await _context.SaveChangesAsync();

            return Ok(new { success = true });
        }

        [HttpDelete("{offeringId}")]
        [ServiceFilter(typeof(AuthFilter))]
        public async Task<IActionResult> DeleteProviderOffer(int offeringId)
        {
            var userId = (int)HttpContext.Items["UserId"]!;

            var provider = await _context.providers.FirstOrDefaultAsync(p => p.UserId == userId);
            if (provider == null)
            {
                return Unauthorized(new { message = "You are not a provider" });
            }

            var offering = await _context.offerings
                .Include(o => o.Provider)
                .FirstOrDefaultAsync(o => o.Id == offeringId);

            if (offering == null)
            {
                return BadRequest(new { message = "Invalid Offering Id" });
            }

            if (offering.Provider.UserId != userId)
            {
                return Unauthorized(new { message = "You are not the owner of this offering" });
            }

            _context.offerings.Remove(offering);
            await _context.SaveChangesAsync();

            return Ok(new { success = true });
        }
    }

    public class FilterRequest
    {
        public List<string>? Job { get; set; }
        public RateFilter? Rate { get; set; }
        public string? Country { get; set; }
        public List<string>? Cities { get; set; }
    }

    public class RateFilter
    {
        public decimal? Min { get; set; }
        public decimal? Max { get; set; }
    }

    public class CreateOfferRequest
    {
        public OfferDto Offer { get; set; } = new();
    }

    public class UpdateOfferRequest
    {
        public OfferDto Offer { get; set; } = new();
    }

    public class OfferDto
    {
        public int ServiceId { get; set; }
        public string Title { get; set; } = string.Empty;
        public decimal? Rate { get; set; }
        public string Curr { get; set; } = string.Empty;
        public bool? Active { get; set; }
    }
}
